metadata/moa/scripts/nbconverted/2.map-broad_id.py

```diff
@@ -0,0 +1,117 @@
+
+
+# In[3]:
+
+
+# 2017 version is missing deprecated_broad_id
+samples_2017['deprecated_broad_id'] = np.nan
+
+
+# In[4]:
+
+
+# The first 13 characters of the Broad ID and the first 14 characters of InChIKey are extracted
+# Year names are appended to column names
+# For some rows InChI are extracted instead of InChIKey -> these are fixed
+
+def id_cleanup(df, year):
+    df.dropna(subset=['InChIKey'], inplace=True)
+    df.reset_index(drop=True, inplace=True)
+    df.InChIKey = df.InChIKey.apply(lambda x: inchi.InchiToInchiKey(x) if (x.startswith('InChI')) else x)
+    df.broad_id = df.broad_id.apply(lambda x: str(x)[:13])
+    df.InChIKey = df.InChIKey.apply(lambda x: str(x)[:14])
+    df = df.drop_duplicates(['InChIKey','pert_iname','broad_id','deprecated_broad_id']).reset_index(drop=True)
+    df = df.rename(columns={'pert_iname':'pert_iname_'+year,
+                            'broad_id':'broad_id_'+year,
+                            'deprecated_broad_id':'deprecated_broad_id_'+year,
+                            'InChIKey':'InChIKey14'})
+    return df
+
+# Grouping samples using InChIKey14 while all other fields are pipe delimited
+
+def group_by_InChIKey14(df, year):
+    df = df.fillna('')
+    df = df.groupby('InChIKey14').agg({'broad_id_'+year : lambda x: '|'.join(np.unique(x)),
+                                       'deprecated_broad_id_'+year: lambda x: merge_dep_id(list(x)),
+                                       'pert_iname_'+year: lambda x: '|'.join(np.unique(x))}).reset_index()
+    return df
+
+# This function cleans up empty depredicated_broad_ids, de-duplicates them and then separates them with pipes
+
+def merge_dep_id(target):
+    joined_target = ('|'.join(target)).split('|')
+    while '' in joined_target:
+        joined_target.remove('')
+    out_target = '|'.join(np.unique(joined_target))
+    return out_target
+
+
+# In[5]:
+
+
+samples_2017 = id_cleanup(samples_2017, "2017")
+samples_2018a = id_cleanup(samples_2018a, "2018a")
+samples_2018b = id_cleanup(samples_2018b, "2018b")
+samples_2020 = id_cleanup(samples_2020, "2020")
+
+samples_2017.head()
+
+
+# In[6]:
+
+
+samples_2017 = group_by_InChIKey14(samples_2017, "2017")
+samples_2018a = group_by_InChIKey14(samples_2018a, "2018a")
+samples_2018b = group_by_InChIKey14(samples_2018b, "2018b")
+samples_2020 = group_by_InChIKey14(samples_2020, "2020")
+
+samples_2017.head()
+
+
+# The four data frames are merged on InChIKey14
+
+# In[7]:
+
+
+merged_df = samples_2017.merge(samples_2018a, on='InChIKey14', how='outer')
+merged_df = merged_df.merge(samples_2018b, on='InChIKey14', how='outer')
+merged_df = merged_df.merge(samples_2020, on='InChIKey14', how='outer')
+
+merged_df= merged_df.replace(to_replace='', value=np.nan)
+
+merged_df.head()
+
+
+# In[8]:
+
+
+merged_df.to_csv('clue/broad_id_map.csv', index=False)
```

Comment:
```suggestion
map_file = os.path.join("clue", "broad_id_map.csv")
merged_df.to_csv(map_file, index=False)
```

It's generally not good practice to hardcode paths. I've also seen people use [`pathlib`](https://docs.python.org/3/library/pathlib.html) before - this may actually be the preferred way now, but I have yet to transition :/